File: November2022/Java/FindMedianFromDataStream.java
// @saorav21994

class MedianFinder {
    
    // Principle : The max element in maxheap should be < than min element in min heap
    // and both the heaps need to be balanced such that |maxheap_size-minheap_size| <= 1

    private PriorityQueue<Integer> minHeap = null;
    private PriorityQueue<Integer> maxHeap = null;    
    
    /** initialize your data structure here. */
    public MedianFinder() {
        minHeap = new PriorityQueue<>();    // Default
        maxHeap = new PriorityQueue<>(Collections.reverseOrder());  // ((a,b)->(b-a))
    }
    
    public void addNum(int num) {
        if (maxHeap.size() == 0 || maxHeap.peek() >= num) {
            maxHeap.offer(num);
        }
        else {
            minHeap.offer(num);
        }
        // Balance the heaps
        if ((maxHeap.size() - minHeap.size()) > 1) {
            int extra = maxHeap.poll();
            minHeap.offer(extra);
        }
        else if ((minHeap.size() - maxHeap.size()) > 1) {
            int extra = minHeap.poll();
            maxHeap.offer(extra);
        }
    }
    
    public double findMedian() {
        if (maxHeap.size() > minHeap.size()) {
            return maxHeap.peek();
        }
        else if (minHeap.size() > maxHeap.size()) {
            return minHeap.peek();
        }
        else {  // both are equal = even = (a+b)/2
            return ((maxHeap.peek() + minHeap.peek())/2.0);
        }
    }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */

File: November2022/C++ Sol/find-median-from-data-stream.cpp
// @Sarang661
class MedianFinder {
private:
    priority_queue<int>q1;
    priority_queue<int,vector<int>,greater<int>>q2;
public:
    MedianFinder() {
        
    }
    
    void addNum(int num) {
      if(q1.size()>0 && q1.top()<=num){
          q2.push(num);
      }
        else if(q2.size()>0 && q2.top()>num){
            q1.push(num);
        }
        else{
            q1.push(num);
        }
    }
    
    double findMedian() {
        while(q1.size()>q2.size()){
            q2.push(q1.top());
            q1.pop();
        }
           while(q2.size()>q1.size()){
            q1.push(q2.top());
            q2.pop();
        }
        double ans = 0.0;
        int size = q1.size()+q2.size();
        if(size%2==0){
            ans = (double)(q1.top()+q2.top())/2.0;
           
            return ans;
        }
        else{
            if(q1.size()>q2.size()){
                return q1.top();
            }
            else{
                return q2.top();
            }
        }
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */